src/verifier.rs

```diff
@@ -78,109 +78,104 @@ pub enum VerifierError {
     #[error("Invalid register specified at instruction {0}")]
     InvalidRegister(usize),
 }
-impl UserDefinedError for VerifierError {}
-impl From<VerifierError> for UserError {
-    fn from(error: VerifierError) -> Self {
-        UserError::VerifierError(error)
-    }
+
+fn adj_insn_ptr(insn_ptr: usize) -> usize {
+    insn_ptr + ebpf::ELF_INSN_DUMP_OFFSET
 }
 
 fn check_prog_len(prog: &[u8]) -> Result<(), VerifierError> {
     if prog.len() % ebpf::INSN_SIZE != 0 {
         return Err(VerifierError::ProgramLengthNotMultiple);
     }
-    if prog.len() > ebpf::PROG_MAX_SIZE {
-        return Err(VerifierError::ProgramTooLarge(prog.len() / ebpf::INSN_SIZE));
-    }
-
     if prog.is_empty() {
         return Err(VerifierError::NoProgram);
     }
-    let last_insn = ebpf::get_insn(prog, (prog.len() / ebpf::INSN_SIZE) - 1);
-    if last_insn.opc != ebpf::EXIT {
-        return Err(VerifierError::InvalidLastInstruction);
-    }
-
     Ok(())
 }
 
 fn check_imm_nonzero(insn: &ebpf::Insn, insn_ptr: usize) -> Result<(), VerifierError> {
     if insn.imm == 0 {
-        return Err(VerifierError::DivisionByZero(insn_ptr));
+        return Err(VerifierError::DivisionByZero(adj_insn_ptr(insn_ptr)));
     }
     Ok(())
 }
 
 fn check_imm_endian(insn: &ebpf::Insn, insn_ptr: usize) -> Result<(), VerifierError> {
     match insn.imm {
         16 | 32 | 64 => Ok(()),
-        _ => Err(VerifierError::UnsupportedLeBeArgument(insn_ptr)),
+        _ => Err(VerifierError::UnsupportedLEBEArgument(adj_insn_ptr(
+            insn_ptr,
+        ))),
     }
 }
 
 fn check_load_dw(prog: &[u8], insn_ptr: usize) -> Result<(), VerifierError> {
-    // We know we can reach next insn since we enforce an EXIT insn at the end of program, while
-    // this function should be called only for LD_DW insn, that cannot be last in program.
+    if insn_ptr + 1 >= (prog.len() / ebpf::INSN_SIZE) {
```

Comment:
Related to the removal of `VerifierError::InvalidLastInstruction` above

Reply:
This should protect against the last instruction being a half `LD_DW` instruction